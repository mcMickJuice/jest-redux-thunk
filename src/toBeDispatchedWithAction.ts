import diff from "jest-diff";
import {
  assertActionTypeProvided,
  assertIsMockFunction,
  toMatchObjectComparison
} from "./utils";
/**
 * ensures that an action was dispatched that matches a certain type
 * for actions that match that type, check each key to see if subset
 * actual dispatched thing
 *
 * similar to "toMatchObject" but with mocks
 */

interface ExpectedAction {
  type: string;
  [prop: string]: any;
}

function toBeDispatchedWithAction(
  this: jest.MatcherUtils,
  receivedDispatchMock: jest.Mock,
  expectedAction: ExpectedAction
) {
  assertIsMockFunction(receivedDispatchMock);
  assertActionTypeProvided(expectedAction.type);

  const calls = receivedDispatchMock.mock.calls;
  // check to see if dispatchMock was called at all
  if (calls.length === 0) {
    return {
      message: () => "Dispatch not called",
      pass: false
    };
  }

  const matchingActionTypeCalls = calls.filter(
    call => call[0].type === expectedAction.type
  );

  if (matchingActionTypeCalls.length === 0) {
    return {
      message: () => `No Action dispatched with type ${expectedAction.type}`,
      pass: false
    };
  }

  // TODO: support this?
  if (matchingActionTypeCalls.length > 1) {
    throw new Error(
      `Action of type ${
        expectedAction.type
      } dispatched multiple times. This matcher only supports action dispatched one time`
    );
  }

  const matchingDispatchedAction = matchingActionTypeCalls[0][0];

  const { type: _, ...otherProps } = expectedAction;

  const pass = Object.keys(otherProps).every(propKey => {
    const actual = matchingDispatchedAction[propKey];
    const expected = expectedAction[propKey];

    return toMatchObjectComparison(actual, expected);
  });

  const message = pass
    ? () =>
        `${this.utils.matcherHint(
          ".not.toBeDispatchedWithActionTypeOrder",
          "dispatch",
          "actionTypeOrder"
        )}\n\nExpected action type sequence not to have matched but it did`
    : () => {
        const diffString = diff(expectedAction, matchingDispatchedAction);

        return `Expected Dispatched Action:\n${this.utils.printExpected(
          expectedAction
        )}\n\nActual Dispatched Action Types (in order):\n${this.utils.printReceived(
          matchingDispatchedAction
        )}\n\nDifference:\n${diffString}`;
      };

  return {
    message,
    pass
  };
}

export default toBeDispatchedWithAction;
