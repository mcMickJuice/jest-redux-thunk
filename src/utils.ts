export const assertIsMockFunction = (mock: any) => {
  if (mock._isMockFunction !== true) {
    throw new Error("Value provided is not a mock function");
  }
};

export const assertActionTypeProvided = (actionType: any) => {
  if (actionType == null) {
    throw new Error("ActionType not provided");
  }
};
