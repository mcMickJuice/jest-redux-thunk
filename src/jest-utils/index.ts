import { equals } from "./jasmineUtils";
import { iterableEquality, subsetEquality } from "./utils";

export const toMatchObjectComparison = (obj1: any, obj2: any): boolean => {
  return equals(obj1, obj2, [iterableEquality, subsetEquality]);
};
