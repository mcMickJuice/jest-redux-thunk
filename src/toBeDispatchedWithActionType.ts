import { assertActionTypeProvided, assertIsMockFunction } from "./utils";

function toBeDispatchedWithActionType(
  this: jest.MatcherUtils,
  receivedDispatchMock: jest.Mock,
  expectedActionType: string
) {
  assertIsMockFunction(receivedDispatchMock);
  assertActionTypeProvided(expectedActionType);

  const calls = receivedDispatchMock.mock.calls;
  // check to see if dispatchMock was called at all
  if (calls.length === 0) {
    return {
      message: () => "Dispatch not called",
      pass: false
    };
  }

  // assume always called with one param
  const pass = calls.some(call => call[0].type === expectedActionType);

  const message = pass
    ? () =>
        `${this.utils.matcherHint(
          ".not.dispatchedWithActionType",
          "dispatch",
          "actionType"
        )} \n\n Expected action type ${this.utils.printExpected(
          expectedActionType
        )} not to have been dispatched but it was`
    : () =>
        `${this.utils.matcherHint(
          ".dispatchedWithActionType",
          "dispatch",
          "actionType"
        )} \n\n Expected action type ${this.utils.printExpected(
          expectedActionType
        )} to have been dispatched but it wasn't`;

  return {
    message,
    pass
  };
}

export default toBeDispatchedWithActionType;
