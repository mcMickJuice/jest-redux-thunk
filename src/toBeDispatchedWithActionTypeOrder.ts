import diff from "jest-diff";
import deepEqual from "lodash.isequal";
import { assertIsMockFunction } from "./utils";

function toBeDispatchedWithActionTypeOrder(
  this: jest.MatcherUtils,
  dispatchMock: jest.Mock,
  expectedActionTypeOrder: string[]
) {
  assertIsMockFunction(dispatchMock);

  const calls = dispatchMock.mock.calls;
  const actualDispatchedActionTypes = calls.map(call => call[0].type);

  const pass = deepEqual(expectedActionTypeOrder, actualDispatchedActionTypes);

  const message = pass
    ? () =>
        `${this.utils.matcherHint(
          ".not.toBeDispatchedWithActionTypeOrder",
          "dispatch",
          "actionTypeOrder"
        )}\n\nExpected action type sequence not to have matched but it did`
    : () => {
        const diffString = diff(
          expectedActionTypeOrder,
          actualDispatchedActionTypes
        );

        return `Expected Dispatched Action Types (in order):\n${this.utils.printExpected(
          expectedActionTypeOrder
        )}\n\nActual Dispatched Action Types (in order):\n${this.utils.printReceived(
          actualDispatchedActionTypes
        )}\n\nDifference:\n${diffString}`;
      };

  return {
    message,
    pass
  };
}

export default toBeDispatchedWithActionTypeOrder;
