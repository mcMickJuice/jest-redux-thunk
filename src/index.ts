import toBeDispatchedWithAction from "./toBeDispatchedWithAction";
import toBeDispatchedWithActionType from "./toBeDispatchedWithActionType";
import toBeDispatchedWithActionTypeOrder from "./toBeDispatchedWithActionTypeOrder";

expect.extend({
  toBeDispatchedWithAction,
  toBeDispatchedWithActionType,
  toBeDispatchedWithActionTypeOrder
});
