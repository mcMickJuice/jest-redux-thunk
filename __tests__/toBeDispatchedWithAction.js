require("../src");

describe("toBeDispatchedWithAction", () => {
  describe("invalid parameters", () => {
    it("fails if no calls are made", () => {
      let error;
      try {
        expect(jest.fn()).toBeDispatchedWithAction({ type: "hello" });
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
    it("fails if dispatchMock is not a mock", () => {
      function fn() {}

      let error;
      try {
        expect(fn).toBeDispatchedWithAction({ type: "hello" });
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
    it("fails if action.type is undefined", () => {
      let error;
      try {
        expect(jest.fn()).toBeDispatchedWithAction({
          payload: {}
        });
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
  });

  describe("affirmative tests", () => {
    it("fails if action.type is never dispatched", () => {
      const dispatchMock = jest.fn();

      dispatchMock({
        type: "NOTHING"
      });

      let error;
      try {
        expect(dispatchMock).toBeDispatchedWithAction({
          type: "SOMETHING"
        });
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
    it("fails if action.type is dispatched multiple times", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock({
        type: expectedActionType
      });

      dispatchMock({
        type: expectedActionType
      });

      let error;
      try {
        expect(dispatchMock).toBeDispatchedWithAction({
          type: expectedActionType
        });
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
    it("fails if action is not subset of actual dispatched action - different payload properties", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });

      let error;

      try {
        expect(dispatchMock).toBeDispatchedWithAction({
          type: expectedActionType,
          payload: {
            name: "mike"
          }
        });
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });

    it("fails if action is not subset of actual dispatched action - payload properties don't match", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });

      let error;

      try {
        expect(dispatchMock).toBeDispatchedWithAction({
          type: expectedActionType,
          payload: {
            items: [1, 2]
          }
        });
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });

    it("passes if action is subset of actual dispatched action - meta check", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock({
        type: expectedActionType,
        meta: {
          message: "hi"
        },
        payload: {
          items: [1, 2, 3]
        }
      });

      expect(dispatchMock).toBeDispatchedWithAction({
        type: expectedActionType,
        meta: {
          message: "hi"
        }
      });
    });

    it("passes if action is subset of actual dispatched action - property subset", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3],
          name: "mike"
        }
      });

      expect(dispatchMock).toBeDispatchedWithAction({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });
    });

    it("passes if multiple actions are dispatched", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock({
        type: "ANOTHER TYPE",
        payload: {
          name: "mike"
        }
      });

      dispatchMock({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });

      expect(dispatchMock).toBeDispatchedWithAction({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });
    });

    it("passes if undefined is dispatched", () => {
      const dispatchMock = jest.fn();
      const expectedActionType = "SOMETHING";

      dispatchMock(undefined);

      dispatchMock({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });

      expect(dispatchMock).toBeDispatchedWithAction({
        type: expectedActionType,
        payload: {
          items: [1, 2, 3]
        }
      });
    });
  });
});
