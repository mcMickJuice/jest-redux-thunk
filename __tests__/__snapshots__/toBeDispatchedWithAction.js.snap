// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeDispatchedWithAction affirmative tests fails if action is not subset of actual dispatched action - different payload properties 1`] = `
"Expected Dispatched Action:
[32m{\\"payload\\": {\\"name\\": \\"mike\\"}, \\"type\\": \\"SOMETHING\\"}[39m

Actual Dispatched Action Types (in order):
[31m{\\"payload\\": {\\"items\\": [1, 2, 3]}, \\"type\\": \\"SOMETHING\\"}[39m

Difference:
[32m- Expected[39m
[31m+ Received[39m

[2m  Object {[22m
[2m    \\"payload\\": Object {[22m
[32m-     \\"name\\": \\"mike\\",[39m
[31m+     \\"items\\": Array [[39m
[31m+       1,[39m
[31m+       2,[39m
[31m+       3,[39m
[31m+     ],[39m
[2m    },[22m
[2m    \\"type\\": \\"SOMETHING\\",[22m
[2m  }[22m"
`;

exports[`toBeDispatchedWithAction affirmative tests fails if action is not subset of actual dispatched action - payload properties don't match 1`] = `
"Expected Dispatched Action:
[32m{\\"payload\\": {\\"items\\": [1, 2]}, \\"type\\": \\"SOMETHING\\"}[39m

Actual Dispatched Action Types (in order):
[31m{\\"payload\\": {\\"items\\": [1, 2, 3]}, \\"type\\": \\"SOMETHING\\"}[39m

Difference:
[32m- Expected[39m
[31m+ Received[39m

[2m  Object {[22m
[2m    \\"payload\\": Object {[22m
[2m      \\"items\\": Array [[22m
[2m        1,[22m
[2m        2,[22m
[31m+       3,[39m
[2m      ],[22m
[2m    },[22m
[2m    \\"type\\": \\"SOMETHING\\",[22m
[2m  }[22m"
`;

exports[`toBeDispatchedWithAction affirmative tests fails if action.type is dispatched multiple times 1`] = `"Action of type SOMETHING dispatched multiple times. This matcher only supports action dispatched one time"`;

exports[`toBeDispatchedWithAction affirmative tests fails if action.type is never dispatched 1`] = `"No Action dispatched with type SOMETHING"`;

exports[`toBeDispatchedWithAction invalid parameters fails if action.type is undefined 1`] = `"ActionType not provided"`;

exports[`toBeDispatchedWithAction invalid parameters fails if dispatchMock is not a mock 1`] = `"Value provided is not a mock function"`;

exports[`toBeDispatchedWithAction invalid parameters fails if no calls are made 1`] = `"Dispatch not called"`;
